File: 979/979.cs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/979.py

using System;
using System.Collections.Generic;

class Program979 {
    static readonly int[][] SUBS_TABLE = new int[][] {
        new int[] {0, 0, 1},
        new int[] {0, 1}
    };

    static (List<int>[] typesList, List<int>[] parent1List, List<int>[] parent2List) BuildLayers(int maxLayer) {
        var typesList = new List<int>[maxLayer + 1];
        var parent1List = new List<int>[maxLayer + 1];
        var parent2List = new List<int>[maxLayer + 1];
        for (int k = 0; k <= maxLayer; k++) {
            typesList[k] = new List<int>();
            parent1List[k] = new List<int>();
            parent2List[k] = new List<int>();
        }

        typesList[0] = new List<int> {0};
        if (maxLayer == 0) {
            return (typesList, parent1List, parent2List);
        }

        for (int i = 0; i < 7; i++) {
            typesList[1].Add(0);
            parent1List[1].Add(0);
            parent2List[1].Add(-1);
        }

        for (int k = 2; k <= maxLayer; k++) {
            var prev = typesList[k - 1];
            int m = prev.Count;
            var cur = new List<int>();
            var p1 = new List<int>();
            var p2 = new List<int>();

            for (int j = 0; j < prev.Count; j++) {
                int t = prev[j];
                int[] block = SUBS_TABLE[t];
                int blen = block.Length;
                for (int pos = 0; pos < blen; pos++) {
                    int ct = block[pos];
                    cur.Add(ct);
                    p1.Add(j);
                    if (pos == blen - 1) {
                        p2.Add((j + 1) % m);
                    } else {
                        p2.Add(-1);
                    }
                }
            }

            typesList[k] = cur;
            parent1List[k] = p1;
            parent2List[k] = p2;
        }

        return (typesList, parent1List, parent2List);
    }

    static (List<int>[] adj, int origin, int[] offsets) BuildBallAdjacency(int maxLayer) {
        var (typesList, parent1List, parent2List) = BuildLayers(maxLayer);
        int[] sizes = new int[maxLayer + 1];
        for (int k = 0; k <= maxLayer; k++) {
            sizes[k] = typesList[k].Count;
        }

        int[] offsets = new int[maxLayer + 1];
        int total = 0;
        for (int k = 0; k <= maxLayer; k++) {
            offsets[k] = total;
            total += sizes[k];
        }

        var adj = new List<int>[total];
        for (int i = 0; i < total; i++) {
            adj[i] = new List<int>();
        }

        void AddEdge(int u, int v) {
            adj[u].Add(v);
            adj[v].Add(u);
        }

        int origin = offsets[0];

        for (int k = 1; k <= maxLayer; k++) {
            int off = offsets[k];
            int m = sizes[k];
            for (int i = 0; i < m; i++) {
                AddEdge(off + i, off + ((i + 1) % m));
            }
        }

        if (maxLayer >= 1) {
            int off1 = offsets[1];
            for (int i = 0; i < sizes[1]; i++) {
                AddEdge(origin, off1 + i);
            }
        }

        for (int k = 2; k <= maxLayer; k++) {
            int off = offsets[k];
            int poff = offsets[k - 1];
            for (int i = 0; i < sizes[k]; i++) {
                AddEdge(off + i, poff + parent1List[k][i]);
                int p = parent2List[k][i];
                if (p != -1) {
                    AddEdge(off + i, poff + p);
                }
            }
        }

        if (maxLayer >= 1) {
            for (int k = 0; k < maxLayer; k++) {
                int off = offsets[k];
                for (int i = 0; i < sizes[k]; i++) {
                    int u = off + i;
                    if (adj[u].Count != 7) {
                        throw new Exception($"Assertion failed: adj degree at ({k},{i}) is {adj[u].Count}");
                    }
                }
            }
        }

        return (adj, origin, offsets);
    }

    static long ComputeF(int n) {
        if (n < 0) return 0L;
        if (n == 0) return 1L;
        int maxLayer = n / 2;
        var (adj, origin, _) = BuildBallAdjacency(maxLayer);
        int nodeCount = adj.Length;
        long[] dp = new long[nodeCount];
        dp[origin] = 1L;

        for (int step = 0; step < n; step++) {
            long[] ndp = new long[nodeCount];
            for (int u = 0; u < nodeCount; u++) {
                long val = dp[u];
                if (val != 0L) {
                    foreach (int v in adj[u]) {
                        ndp[v] += val;
                    }
                }
            }
            dp = ndp;
        }

        return dp[origin];
    }

    static void Main(string[] args) {
        long f4 = ComputeF(4);
        if (f4 != 119L) {
            throw new Exception($"Assertion failed: F(4) == {f4}, expected 119");
        }
        Console.WriteLine(ComputeF(20));
    }
}

File: 979/979.rs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/979.py

const SUBS_TABLE: &[&[i32]] = &[
    &[0, 0, 1],
    &[0, 1],
];

fn build_layers(max_layer: usize) -> (Vec<Vec<i32>>, Vec<Vec<i32>>, Vec<Vec<i32>>) {
    let mut types_list: Vec<Vec<i32>> = vec![vec![]; max_layer + 1];
    let mut parent1_list: Vec<Vec<i32>> = vec![vec![]; max_layer + 1];
    let mut parent2_list: Vec<Vec<i32>> = vec![vec![]; max_layer + 1];

    types_list[0] = vec![0];
    if max_layer == 0 {
        return (types_list, parent1_list, parent2_list);
    }

    types_list[1] = vec![0; 7];
    parent1_list[1] = vec![0; 7];
    parent2_list[1] = vec![-1; 7];

    for k in 2..=max_layer {
        let prev = types_list[k - 1].clone();
        let m = prev.len();
        let mut cur: Vec<i32> = vec![];
        let mut p1: Vec<i32> = vec![];
        let mut p2: Vec<i32> = vec![];

        for (j, &t) in prev.iter().enumerate() {
            let block = SUBS_TABLE[t as usize];
            let blen = block.len();
            for (pos, &ct) in block.iter().enumerate() {
                cur.push(ct);
                p1.push(j as i32);
                if pos == blen - 1 {
                    p2.push(((j + 1) % m) as i32);
                } else {
                    p2.push(-1);
                }
            }
        }

        types_list[k] = cur;
        parent1_list[k] = p1;
        parent2_list[k] = p2;
    }

    (types_list, parent1_list, parent2_list)
}

fn build_ball_adjacency(max_layer: usize) -> (Vec<Vec<usize>>, usize, Vec<usize>) {
    let (types_list, parent1_list, parent2_list) = build_layers(max_layer);
    let sizes: Vec<usize> = (0..=max_layer).map(|k| types_list[k].len()).collect();

    let mut offsets = vec![0usize; max_layer + 1];
    let mut total = 0usize;
    for k in 0..=max_layer {
        offsets[k] = total;
        total += sizes[k];
    }

    let mut adj: Vec<Vec<usize>> = vec![vec![]; total];

    let mut add_edge = |adj: &mut Vec<Vec<usize>>, u: usize, v: usize| {
        adj[u].push(v);
        adj[v].push(u);
    };

    let origin = offsets[0];

    for k in 1..=max_layer {
        let off = offsets[k];
        let m = sizes[k];
        for i in 0..m {
            add_edge(&mut adj, off + i, off + ((i + 1) % m));
        }
    }

    if max_layer >= 1 {
        let off1 = offsets[1];
        for i in 0..sizes[1] {
            add_edge(&mut adj, origin, off1 + i);
        }
    }

    for k in 2..=max_layer {
        let off = offsets[k];
        let poff = offsets[k - 1];
        for i in 0..sizes[k] {
            let par1 = parent1_list[k][i] as usize;
            add_edge(&mut adj, off + i, poff + par1);
            let p = parent2_list[k][i];
            if p != -1 {
                add_edge(&mut adj, off + i, poff + p as usize);
            }
        }
    }

    if max_layer >= 1 {
        for k in 0..max_layer {
            let off = offsets[k];
            for i in 0..sizes[k] {
                let u = off + i;
                assert_eq!(
                    adj[u].len(),
                    7,
                    "Assertion failed: adj degree at ({},{}) is {}",
                    k,
                    i,
                    adj[u].len()
                );
            }
        }
    }

    (adj, origin, offsets)
}

fn compute_f(n: i64) -> i64 {
    if n < 0 {
        return 0;
    }
    if n == 0 {
        return 1;
    }
    let max_layer = (n / 2) as usize;
    let (adj, origin, _) = build_ball_adjacency(max_layer);
    let node_count = adj.len();
    let mut dp: Vec<i64> = vec![0; node_count];
    dp[origin] = 1;

    for _ in 0..n {
        let mut ndp: Vec<i64> = vec![0; node_count];
        for u in 0..node_count {
            let val = dp[u];
            if val != 0 {
                for &v in &adj[u] {
                    ndp[v] += val;
                }
            }
        }
        dp = ndp;
    }

    dp[origin]
}

fn main() {
    let f4 = compute_f(4);
    assert_eq!(f4, 119, "Assertion failed: F(4) == {}, expected 119", f4);
    println!("{}", compute_f(20));
}
